apiserver/encrypting-data-at-datastore-layer.md

```diff
@@ -0,0 +1,243 @@
+ }
+```
+
+The default value of empty string for the type will imply `identity` (meaning no encryption of the datastore layer).
+
+The kube api server operator and openshift api server operator will observe this config and begin encrypting sensitive resources when it is set.  The operators are responsible for determining what resources to encrypt.
+
+### User Stories
+
+#### Story 1
+
+User enables encryption by setting `apiserver.spec.encryption.type` to `aescbc`.  After some time passes, user makes a backup of etcd.  The user confirms that the secret values are encrypted by checking to see if they have the `k8s:enc:aescbc:v1:` prefix.
+
+### Implementation Details/Notes/Constraints
+
+Key management is per resources (GroupResources). Different resources that are encrypted have different keys. Keys per resource are numbered with strict increasing unsigned integers. Different resources can have different or overlapping key numbers at the same time.
+
+etcd encryption keys are created and stored in secrets, by convention they are named as:
+
+    <component>-<group>-<resource>-encryption-<unsigned integer #>
+
+with the mandatory and authoritative labels:
+
+- `encryption.operator.openshift.io/component`
+- `encryption.operator.openshift.io/group`
+- `encryption.operator.openshift.io/resource`
+
+##### Key Life Cycle
+
+A key known to the cluster is in one of the following states:
+
+1. **created** - the key secret is created
+1. **read-key** - all API servers have observed and configured the key as a read key
+1. **write-key** - all API servers have observed and configured the key as a write key
+1. **migrated** - since the key has been write-key, all key-values in etcd have been rewritten (=migrated)
+1. **deleted** - all API servers have removed the key as read key
+
+They go through 1-4 via these transitions:
+
+- 1 -> 2: the key is deployed via new API server revisions rolled out to all master nodes.
+- 2 -> 3: the key is deployed via new API server revisions rolled out to all master nodes.
+- 3 -> 4: a migration mechanism rewrites all key-values in etcd.
+
+After another key has been migrated to, i.e. reaches 4, an old key can be decommissioned:
+- 4 -> 5: the old key secret in the cluster is deleted and asynchronously all API servers are updated on the master nodes (no need to wait for that to finish as the read-key is not used anymore after 4).
+
+##### Resources
+
+This list of resources is not configurable.  The following resources are encrypted:
+
+1. secrets
+1. config maps
+1. routes
+1. oauth access tokens
+1. oauth authorize tokens
+
+##### Controllers
+
+The encryption key rotation logic will be implemented using a distributed state machine with a series of controllers that each perform a single, _simple_ task:
+
+1. A controller that looks at all keys in use and generates a single complete encryption config secret in `openshift-config-managed` (this matches the encryption config file format used in upstream)
```

Comment:
I know that you did this because you didn't want to have sync based on a name pattern, but explain and defend that.